students/patchcarrier/Lesson04/trigram.py

```diff
@@ -0,0 +1,217 @@
+        word_count = len(line_i)
+        
+        # The new is the two previous words
+        new_key = ' '.join(line_i[-2:])
+        
+        while (new_key in tri_d) and (word_count < word_limit):
+            
+            # get list value for the given key
+            newword_list = tri_d[new_key]
+            # the new word is randomly selected from the words in the list
+            # if multiple are present
+            new_word = newword_list[randint(0, len(newword_list) - 1)]
+            last_word = line_i[-1]
+            word_count += 1
+            
+            if char_count + len(new_word) > max_nchars:
+                outfile.write(' '.join(line_i))
+                outfile.write('\n')
+                # reset line_i and char_count
+                line_i = [new_word]
+                char_count = len(new_word)  
+            else:
+                line_i.append(new_word)
+                char_count += len(new_word)
+                
+            new_key = ' '.join([last_word, new_word])
+                
+        # save the current results of line_i
+        last_line = line_i.copy()
+        
+        # If the output doesn't end with a period, continue looking for 
+        # more words until a word with a period is found. If the total number of 
+        # words exceeds 120 percent of word_limit, just otput the results
+        # ignoring this search for a period
+        
+        # list of word lists for the additional lines 
+        lines_after_wordlim =[line_i]
+        count = 0
+        while (new_key in tri_d) and (line_i[-1][-1] != '.'):
+            count += 1
+            print(count)
+            if word_count > absolute_wordlim: 
+                print('broke out')
+                break
+            
+            # get list value for the given key
+            newword_list = tri_d[new_key]
+            # the new word is randomly selected from the words in the list
+            # if multiple are present
+            new_word = newword_list[randint(0, len(newword_list) - 1)]
+            last_word = lines_after_wordlim[-1][-1]
+            word_count += 1
+            
+            if char_count + len(new_word) > max_nchars:
+                lines_after_wordlim.append([new_word])
+                char_count = len(new_word)  
+            else:
+                lines_after_wordlim[-1].append(new_word)
+                char_count += len(new_word)
+                
+            new_key = ' '.join([last_word, new_word])
+            
+        # The search for a period was unsuccessful, ingore the additional lines and
+        # output what the last line would have been before the search
+        if line_i[-1][-1] != '.':
+            outfile.write(' '.join(last_line))      
+        # the search was successful, append the additional lines that end with
+        # a period
+        else:
+            
+            for line in lines_after_wordlim:
+                outfile.write(' '.join(line))
+            return
```

Comment:
for readability, above while loop is quite long with a lot of logic inside - recommend breaking code like that into some logical functions that can improve the readability.